File: python/performances/string_building.py
"""문자열 누적: += 반복 vs ''.join

- 동작: 여러 문자열 조각을 하나의 문자열로 합치기
- 차이: += 반복은 누적 복사로 O(N^2)까지 커질 수 있고,
       join은 한 번에 할당하여 O(N)
"""

from __future__ import annotations

import string
import time


def build_with_plus(chunks: list[str]) -> str:
    result = ""
    for ch in chunks:
        result += ch
    return result


def build_with_join(chunks: list[str]) -> str:
    return "".join(chunks)


def main() -> None:
    chunks = list(string.ascii_letters) * 80_000

    s = time.perf_counter()
    a = build_with_plus(chunks)
    t1 = time.perf_counter() - s

    s = time.perf_counter()
    b = build_with_join(chunks)
    t2 = time.perf_counter() - s

    print(f"+= 길이={len(a)}, 시간={t1:.4f}s")
    print(f"join 길이={len(b)}, 시간={t2:.4f}s")


if __name__ == "__main__":
    main()

File: javascript/performances/string_building.js
const letters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ';
const chunks = Array.from({ length: letters.length * 80000 }, (_, i) => letters[i % letters.length]);

function buildWithPlus(parts) {
  let result = '';
  for (const ch of parts) result += ch;
  return result;
}

function buildWithJoin(parts) {
  return parts.join('');
}

let s = process.hrtime.bigint();
const a = buildWithPlus(chunks);
let t1 = Number(process.hrtime.bigint() - s) / 1e9;

s = process.hrtime.bigint();
const b = buildWithJoin(chunks);
let t2 = Number(process.hrtime.bigint() - s) / 1e9;

console.log(`+= 길이=${a.length}, 시간=${t1.toFixed(4)}s`);
console.log(`join 길이=${b.length}, 시간=${t2.toFixed(4)}s`);
